pkg/notifications/fetch.go

```diff
@@ -0,0 +1,159 @@
+package notifications
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/icinga/icinga-go-library/backoff"
+	"github.com/icinga/icinga-go-library/retry"
+	"github.com/icinga/icinga-go-library/types"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+	"time"
+)
+
+// fetchHostServiceFromRedis retrieves the host and service names from Redis.
+//
+// If serviceId is nil, only the host name is fetched. Otherwise, both host and service name is fetched.
+func (client *Client) fetchHostServiceFromRedis(
+	ctx context.Context,
+	hostId, serviceId types.Binary,
+) (hostName string, serviceName string, err error) {
+	getNameFromRedis := func(ctx context.Context, typ, id string) (string, error) {
+		key := "icinga:" + typ
+
+		var data string
+		if err := retry.WithBackoff(
+			ctx,
+			func(ctx context.Context) (err error) {
+				data, err = client.redisClient.HGet(ctx, key, id).Result()
+				return
+			},
+			retry.Retryable,
+			backoff.DefaultBackoff,
+			retry.Settings{},
+		); err != nil {
+			return "", errors.Wrapf(err, "redis HGET %q, %q failed", key, id)
+		}
+
+		var result struct {
+			Name string `json:"name"`
+		}
+		if err := json.Unmarshal([]byte(data), &result); err != nil {
+			return "", errors.Wrap(err, "failed to unmarshal redis result")
+		}
+
+		return result.Name, nil
+	}
+
+	hostName, err = getNameFromRedis(ctx, "host", hostId.String())
+	if err != nil {
+		return
+	}
+
+	if serviceId != nil {
+		serviceName, err = getNameFromRedis(ctx, "service", serviceId.String())
+		if err != nil {
+			return
+		}
+	}
+
+	return
+}
+
+// fetchCustomVarFromSql retrieves custom variables for the hsot and service from SQL.
+//
+// If serviceId is nil, only the host custom vars are fetched. Otherwise, both host and service custom vars are fetched.
+func (client *Client) fetchCustomVarFromSql(
+	ctx context.Context,
+	hostId, serviceId types.Binary,
+) (map[string]string, error) {
+	type customVar struct {
+		Name  string `db:"name"`
+		Value string `db:"value"`
+	}
+
+	getCustomVarsFromSql := func(ctx context.Context, typ string, id types.Binary) ([]customVar, error) {
+		stmt, err := client.db.Preparex(client.db.Rebind(
+			`SELECT customvar.name AS name, customvar.value AS value
+			FROM ` + typ + `_customvar
+			LEFT JOIN customvar
```

Comment:
I'm not sure about the **`LEFT`** `JOIN` here. Does it serve a purpose? I think it will have an effect but I'm unsure if that's actually intentional: In case there's no row to join, this will have `NULL` values in the result. Given that the members of `customVar` are of type `string`, that should then result in an error.